### PR TITLE
make cargo submodule optional

### PR DESCRIPTION
**Pull request by onur-ozkan**
Right now, we fetch the cargo submodule no matter what, even if the command we are running doesn't need it (e.g., `x build compiler library`). This PR changes that to only fetch the cargo submodule when it's necessary.

For more context, see the zulip thread: https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/Why.20is.20cargo.20always.20checked.20out.3F

### CODE REVIEW

**Comment by onur-ozkan** [comment]
I will wait until #126230 is merged so I don't need to duplicate the tidy part.

File: src/bootstrap/src/core/build_steps/doc.rs [hunk 1]
```diff
@@ -921,6 +920,15 @@ macro_rules! tool_doc {
             /// we do not merge it with the other documentation from std, test and
             /// proc_macros. This is largely just a wrapper around `cargo doc`.
             fn run(self, builder: &Builder<'_>) {
+                let source_type = SourceType::InTree;
+                $(
+                    let _ = source_type; // silent the "unused variable" warning
```
Comment (RalfJung) [inline src/bootstrap/src/core/build_steps/doc.rs:925]:
```suggestion
                    let _ = source_type; // silence the "unused variable" warning
```

**Review by ehuss — COMMENTED** [review]
Looks like the doc comment for `workspace_members` should be updated.

I think the checkout is also needed in `tool::Cargo`, otherwise commands like `./x.py test src/tools cargo` or `./x.py build src/tools/cargo` will fail.

It will also need to be added to the vendor command to support `./x.py vendor`.

**Comment by onur-ozkan** [comment]
Rebased.

@rustbot ready

File: src/bootstrap/src/core/metadata.rs [hunk 2]
```diff
@@ -89,12 +86,8 @@ fn workspace_members(build: &Build) -> impl Iterator<Item = Package> {
 
     // Collects `metadata.packages` from all workspaces.
     let packages = collect_metadata("Cargo.toml");
-    let cargo_packages = collect_metadata("src/tools/cargo/Cargo.toml");
     let ra_packages = collect_metadata("src/tools/rust-analyzer/Cargo.toml");
```
Comment (Kobzol) [inline src/bootstrap/src/core/metadata.rs:89]: While we're at it, shouldn't we also remove Rust analyzer? It also isn't a member of the workspace, right?

File: src/bootstrap/src/core/build_steps/doc.rs [hunk 3]
```diff
@@ -888,12 +888,11 @@ impl Step for Rustc {
 macro_rules! tool_doc {
     (
         $tool: ident,
-        $should_run: literal,
         $path: literal,
         $(rustc_tool = $rustc_tool:literal, )?
-        $(in_tree = $in_tree:literal ,)?
         $(is_library = $is_library:expr,)?
         $(crates = $crates:expr)?
+        $(, submodule $(= $submodule:literal)? )?
```
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/doc.rs:895]: Why isn't this just an optional argument, rather than a repeating argument?

**Review by Kobzol — COMMENTED** [review]
Tried all commands that I could think of and everything seemed to work.

File: src/bootstrap/src/core/build_steps/doc.rs (hunk 3 above)
Comment (onur-ozkan) [reply src/bootstrap/src/core/build_steps/doc.rs:895]: It's already an optional argument, not a repeating one

Comment (Kobzol) [reply src/bootstrap/src/core/build_steps/doc.rs:895]: Sorry :man_facepalming: I misread the macro.

**Comment by Kobzol** [comment]
Ok, looks good! It's possible that we have missed something, but the main commands all seem to work, so hopefully it should be fine (if not, we know what to do, sprinkle `cargo` checkout on more places).

Feel free to r=me once CI is green.

**Pull request merged**